File: 20-valid-parentheses/20-valid-parentheses.scala
import scala.collection.immutable.HashMap
import scala.collection.mutable.Stack
object Solution {
  def isValid(s: String): Boolean = {
    var map = HashMap('(' -> ')', '[' -> ']', '{' -> '}')
    var stack = new Stack[Char]
    for (ch <- s) {
      if (map contains ch) {
        stack.push(map(ch))
      } else {
        if (stack.isEmpty || stack.pop != ch)
          return false
      }
    }
    return stack.isEmpty
  }
}

File: 20-valid-parentheses/20-valid-parentheses.go
func isValid(s string) bool {
  stack := make([]rune, 0, len(s))
  for _, ch := range s {
    switch ch {
      case '(': stack = append(stack, ')')
      case '[': stack = append(stack, ']')
      case '{': stack = append(stack, '}')
      default:
        if len(stack) == 0 || ch != stack[len(stack)-1] {
          return false
        } else {
          stack = stack[:len(stack)-1]
        }
    }
  }
  return len(stack) == 0
}